### ISSUE TITLE
Unmarshaling data using a struct with the jose.JSONWebKeySet struct embedded stopped working on v4.0.3

### ISSUE THREAD

**Issue by amartinezfayo** [go-jose/go-jose#136]
We are trying to update go-jose from 4.0.2 to 4.0.3 in the [SPIRE](https://github.com/spiffe/spire/pull/5288) and [go-spiffe](https://github.com/spiffe/go-spiffe/pull/292) projects, but unit tests are failing with 4.0.3.
Looking at #130, it can be seen the [introduction of the UnmarshalJSON method in the JSONWebKeySet type](https://github.com/go-jose/go-jose/blob/main/jwk.go#L365), which influences the way that a `json.Unmarshal` works when unmarshaling data using a struct with the `jose.JSONWebKeySet` type embedded. This is because now it can be called during the decoding process, as the decoder [now encounters this Unmarshaler that can be called](https://github.com/golang/go/blob/master/src/encoding/json/decode.go#L479) but the result is that [the struct that holds the `jose.JSONWebKeySet` struct](https://github.com/spiffe/spire/blob/main/pkg/common/bundleutil/types.go#L12-L16) does not get the fields properly populated (e.g. the RefreshHint field).
I'm not quite sure if we are doing something wrong in our side. To see if this change is related with the failure, I deleted this function from the 4.0.3 code and the parsing / decoding works as expected, with the tests passing.
I would really appreciate some help to find a solution here. Thanks!

**Comment by jsha** [comment]
Hm, that's an interesting use case. I see you have this type:

```go
type bundleDoc struct {
	jose.JSONWebKeySet
	SequenceNumber *uint64 `json:"spiffe_sequence,omitempty"`
	RefreshHint    *int64  `json:"spiffe_refresh_hint,omitempty"`
}
```

And you're saying that if you parse, e.g. https://github.com/spiffe/go-spiffe/blob/94335b268274926928c08fc5ce21f12a4e1ea044/v2/bundle/spiffebundle/testdata/spiffebundle_valid_2.json#L62-L63 into it, the SequenceNumber and RefreshHint fields do not get populated?

It seems like this is some surprising behavior of Go's `encoding/json` in how it unmarshals into structs with embedded structs in them. https://pkg.go.dev/encoding/json

> To unmarshal JSON into a value implementing [Unmarshaler](https://pkg.go.dev/encoding/json#Unmarshaler), Unmarshal calls that value's [Unmarshaler.UnmarshalJSON] method,

The surprising interaction here is that `bundleDoc` implements `Unmarshaler`, because it has the `UnmarshalJSON` function, because it embeds a struct that has that function. And of course its implementation of `UnmarshalJSON` is the one on `JSONWebKeySet`, which only knows about the JSONWebKeySet and doesn't even have a way to assign fields on the parent struct.

This does break running code like yours, so I'm inclined to revert and publish a new release, and possibly retract 4.0.3. The new release would be breaking with regards to 4.0.3. Anyone who already started relying on JSONWebKeySet to accept unrecognized keys would start getting an error. However, between the two types of breakage I think the one where you get an error is preferable to the one that could be silently missed by anyone without solid tests. Kudos to spire and spiffe for having those tests, and thanks for letting us know.

**Comment by jsha** [comment]
Fixed in v4.0.4: https://github.com/go-jose/go-jose/releases/tag/v4.0.4

**Comment by amartinezfayo** [comment]
Thank you so much for the quick turnaround!
I was able to confirm that this is fixed in the 4.0.4 release, with our tests passing. Thanks again @jsha and @mcpherrinm.
